src/Illuminate/Database/Schema/Grammars/MySqlGrammar.php

```diff
@@ -229,6 +229,24 @@ public function compileUnique(Blueprint $blueprint, Fluent $command)
         return $this->compileKey($blueprint, $command, 'unique');
     }
 
+    /**
+     * Compile a unique index ignoring a null column command.
+     *
+     * @param  \Illuminate\Database\Schema\Blueprint  $blueprint
+     * @param  \Illuminate\Support\Fluent  $command
+     * @return string
+     */
+    public function compileUniqueIgnoreNull(Blueprint $blueprint, Fluent $command)
+    {
+        return sprintf('create unique index %s on %s%s (%s, %s)',
+            $this->wrap($command->index),
+            $this->wrapTable($blueprint),
+            $command->algorithm ? ' using '.$command->algorithm : '',
+            $this->columnize($command->columns),
+            "(CASE WHEN {$this->wrap($command->nullColumn)} IS NULL THEN 0 ELSE {$this->wrap($command->nullColumn)} END)",
```

Comment:
The `0` is any value for the column that will not be treated as unique when doing comparisons. 0 is good for ints / timestamps etc.. but might not be suitable for all scenarios